### ISSUE TITLE
Add a patcher to handle final classes

### ISSUE THREAD

**Issue by Th3Mouk** [kahlan/kahlan#327]
Hi !

Thanks for this super testing framework.

Currently I can't `Double::instance` on a class, after I instanciate it myself, in a previous specification (following the test suite order).

I tried to debug and fix it myself, but I don't have enough background.
I'm blocking here: https://github.com/kahlan/kahlan/blob/b0341a59cefd2697f6e104bbcf045ee36a189ea2/src/Plugin/Double.php#L97
IMO classes must be destroyed after each block to improve isolation.

I think an example worth than a long discussion.
I successfully reproduce the bug on a very simple repo.

Notice the subterfuge with `A` and `B` folders, because if specification order is reversed, the test suite is ✅.

https://github.com/Th3Mouk/debug-kahlan

You just have to clone and launch `make reproduce` command.

PS: I was wondering how I can contribute to improve documentation ?

Thanks for your time

**Comment by jails** [comment]
Glad you like it :+1: 
A double on a class must follow the class constructor signature. Since your `EvilClass` explicitly require an argument for the constructor method, you should pass an argument to make it work like so:
```php
        $dummy = \Kahlan\Plugin\Double::instance([
            'class' => \Debug\EvilClass::class,
            'args' => ['something']
        ]);
```
The documentation has it's [own repository](https://github.com/kahlan/docs) so feel free to do some PR !

**Comment by Th3Mouk** [comment]
I understand but this is only a simple example.

My real problem occurs in a class with 5 objects arguments, but I don't want to make a dummy of each, because it's a tree dependency of doom.

First argument take 3 object arguments etc ...

I always made like this, 

```php
$stub = \Kahlan\Plugin\Double::instance([
            'class' => \Debug\EvilClass::class,
            'methods' => ['__construct']
        ]);
```

But this is the first time I met this weird case, were it doesn't work.

**Comment by Th3Mouk** [comment]
I've made another example on a branch where test suite is passing.

https://github.com/Th3Mouk/debug-kahlan/tree/extends

I remove the `final` keyword on `EvilClass` and use the `extends` option instead of `class`.
Now it work. But my project use dependencies were classes are final.
So the solution was to use `class` option, before this issue. I can't fix my problem this way.

Notice that if you remove this line: https://github.com/Th3Mouk/debug-kahlan/blob/extends/spec/B/EvilBug.spec.php#L7
The test suite fail, with exactly the same error message on the master branch.

So i think that the `methods` option just not working properly on the master branch.

**Comment by jails** [comment]
> IMO classes must be destroyed after each block to improve isolation.

Oh ok I see, indeed in PHP created classes can't be destroyed, once created it lives forever. That's why the `'class'` parameter shouldn't ever be used, since it hardcodes the name of the double. The working pattern to stub the constructor is:
```php
$stub = \Kahlan\Plugin\Double::instance([
    'extends' => \Debug\EvilClass::class,
    'methods' => ['__construct']
]);
```
However this, of course, doesn't work for `final` classes. The `final` keyword gained a lot of popularity lately but for bad reasons imo. It's just a trouble maker, and this kind of issue is just an example of the troubles it can make.

Maybe the best option could be to create a new Patcher to remove the `final` keywords for some classes.

**Comment by Th3Mouk** [comment]
Ok thanks for the clue.

I'm not totally agree concerning final classes but it's another debate 😄.

I will have a look, if i can produce something myself but I lack of time.

**Comment by ddziaduch** [comment]
I recently started using Final Classes. I had also troubles testing them. So what I did? I pulled the Interfaces and started using them as dependencies. They are now injected instead of final classes. That solved all my issues. Final classes are good 🙂

**Comment by Th3Mouk** [comment]
I currently use DDD approach so final classes are ok for port/adapters but I can't apply this on my Domain. Interfaces doesn't make sens on Aggregates, value objects and more.

**Comment by ddziaduch** [comment]
@Th3Mouk why don't you use just true object? If they are value object, they do not have any logic inside, so I assume they are safe to use in test env :)

**Comment by Th3Mouk** [comment]
😄I prefer to test my domain, I already found some weird behavior, and value object have validation logic. Ex: string not empty, pattern matching.

**Comment by jails** [comment]
Arbitrary constraints you can define at a syntax level are as relevant as the developer is skilled.
Maybe having Kahlan to dynamically remove the `final` keyword through a Just In Time patcher would be the easy way here. 🙂

**Comment by jails** [comment]
Fixed in Kahkan 4.7
